webapp/packages/plugin-navigation-tree/src/TreeNew/useTreeData.ts

```diff
@@ -0,0 +1,135 @@
+  rootId: string;
+
+  getNode(id: string): INode;
+  getChildren: (node: string) => string[];
+  getState(id: string): Readonly<INodeState>;
+
+  updateState(id: string, state: Partial<INodeState>): void;
+  load(nodeId: string, manual: boolean): Promise<void>;
+  update(): Promise<void>;
+}
+
+export function useTreeData(options: IOptions): ITreeData {
+  options = useObservableRef(
+    {
+      ...options,
+      childrenTransformers: [...(options.childrenTransformers || [])],
+      nodeTransformers: [...(options.nodeTransformers || [])],
+      stateTransformers: [...(options.stateTransformers || [])],
+    },
+    {
+      rootId: observable.ref,
+      getNode: observable.ref,
+      getChildren: observable.ref,
+      load: observable.ref,
+
+      childrenTransformers: observable.ref,
+      nodeTransformers: observable.ref,
+      stateTransformers: observable.ref,
+    },
+  );
+  const state = useTreeState(options.externalState);
+  const [nodeCache] = useState(
+    () =>
+      new MetadataMap<string, IComputedValue<INode>>(id =>
+        computed(() => applyTransforms(id, options.getNode(id), [rootNodeTransformer(options.rootId), ...(options.nodeTransformers || [])])),
+      ),
+  );
+  const [childrenCache] = useState(
+    () =>
+      new MetadataMap<string, IComputedValue<string[]>>(id =>
+        computed(() => applyTransforms(id, options.getChildren(id), options.childrenTransformers)),
+      ),
+  );
+  const [stateCache] = useState(
+    () =>
+      new MetadataMap<string, IComputedValue<INodeState>>(id =>
+        computed(() => applyTransforms(id, state.getState(id), [rootNodeStateTransformer(options.rootId), ...(options.stateTransformers || [])])),
+      ),
+  );
+
+  const treeData = useObservableRef(
+    () => ({
+      getNode(id: string): INode {
+        return nodeCache.get(id).get();
+      },
+      getChildren(nodeId: string): string[] {
+        return childrenCache.get(nodeId).get();
+      },
+      getState(id: string): Readonly<INodeState> {
+        return stateCache.get(id).get();
+      },
+      updateState(id: string, state: Partial<INodeState>) {
+        this.state.updateState(id, state);
+      },
+      async load(nodeId: string, manual: boolean) {
+        await options.load(nodeId, manual);
+      },
+      async update() {
+        const nodes = [this.rootId];
+
+        while (nodes.length > 0) {
+          const nodeId = nodes.shift()!;
+          const state = this.state.getState(nodeId);
+
+          if (!state.expanded) {
+            continue;
+          }
+
+          await options.load(nodeId, false);
+
+          const children = this.getChildren(nodeId);
+
+          if (children.length === 0) {
+            this.state.updateState(nodeId, { expanded: false });
+            continue;
+          }
+
+          nodes.push(...children);
+        }
+      },
+    }),
+    { state: observable.ref, rootId: observable.ref },
+    { state, rootId: options.rootId },
```

Comment:
this is not a pure function. please add here some deps so all functions uses `this` operator instead of get what it need outside of this scope